### ISSUE TITLE
Big crash, unable to find the source of bug

### ISSUE THREAD

**Issue by ldidry** [ether/etherpad-lite#2669]
Today, one of our etherpad (https://mensuel.framapad.org) starts crashing again and again and again, with no helpful trace (not helpful to me).

    [2015-05-19 16:26:03.180] [ERROR] console - ^{[39mTypeError: Cannot set property 'timestamp' of null
    at /var/www/mensuel.framapad/src/node_modules/ueberDB/CacheAndBufferLayer.js:317:41
    at fn (/var/www/mensuel.framapad/src/node_modules/ueberDB/node_modules/async/lib/async.js:525:34)
    at /var/www/mensuel.framapad/src/node_modules/ueberDB/node_modules/async/lib/async.js:449:34
    at process._tickDomainCallback (node.js:463:13)

Seems really close to issue #2454

Any idea for a fix ?

Informations:
* Etherpad version: 1.5.6
* NodeJs version: v0.10.29
* MongoDB version: 2.4.10

**Comment by JohnMcLear** [comment]
Update Node to 12+

**Comment by ldidry** [comment]
Upgrading to node 12+ did the job, without checkouting develop branch, thanks!

Do I really need to checkout develop branch? (We have 8 different etherpad instances on the same server and I don't really want or have time to checkout all the instances right now)

**Comment by JohnMcLear** [comment]
It's up to you how you manage your systems dude!

**Comment by ldidry** [comment]
Well, I thinked you say to checkout develop branch because of incompatibilities, and I wanted to be sure it was needed. As it seems it's not mandatory, I will not.

Anyway, big thx for your really quick and working answer! You're my etherpad hero ! :blush:

**Comment by ldidry** [comment]
Hum. I just got the `Cannot set property 'colorId' of null` from #2454, so I checked out develop branch, and now I got:

    TypeError: Cannot set property 'writingInProgress' of undefined
        at /var/www/mensuel.framapad/src/node_modules/ueberDB/CacheAndBufferLayer.js:488:56
        at /var/www/mensuel.framapad/src/node_modules/ueberDB/mongodb_db.js:150:29
        at /var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/collection/core.js:716:7
        at /var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/collection/core.js:640:7
        at Server.Base._callHandler (/var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/connection/base.js:453:41)
        at /var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/connection/server.js:487:18
        at MongoReply.parseBody (/var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/responses/mongo_reply.js:68:5)
        at null.<anonymous> (/var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/connection/server.js:445:20)
        at emit (events.js:107:17)
        at null.<anonymous> (/var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/connection/connection_pool.js:207:13)

Any idea ?

* Etherpad version: develop
* NodeJs version: v0.12.3
* MongoDB version: 2.4.10

**Comment by JohnMcLear** [comment]
Sounds like a mongodb issue then

On 20 May 2015 14:10, Luc Didry <notifications@github.com> wrote:

Hum. I just got the Cannot set property 'colorId' of null from #2454<https://github.com/ether/etherpad-lite/issues/2454>, so I checked out develop branch, and now I got:

TypeError: Cannot set property 'writingInProgress' of undefined
    at /var/www/mensuel.framapad/src/node_modules/ueberDB/CacheAndBufferLayer.js:488:56
    at /var/www/mensuel.framapad/src/node_modules/ueberDB/mongodb_db.js:150:29
    at /var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/collection/core.js:716:7
    at /var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/collection/core.js:640:7
    at Server.Base._callHandler (/var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/connection/base.js:453:41)
    at /var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/connection/server.js:487:18
    at MongoReply.parseBody (/var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/responses/mongo_reply.js:68:5)
    at null.<anonymous> (/var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/connection/server.js:445:20)
    at emit (events.js:107:17)
    at null.<anonymous> (/var/www/mensuel.framapad/node_modules/mongodb/lib/mongodb/connection/connection_pool.js:207:13)


Any idea ?

  *   Etherpad version: develop
  *   NodeJs version: v0.12.3
  *   MongoDB version: 2.4.10

**Comment by ldidry** [comment]
Maybe UeberDB could check that the key exists, no?

In  `ueberDB/CacheAndBufferLayer.js`, line 488:

      //set the writingInProgress flag to false
      for(var i in operations)
      {
        db.buffer[operations[i].key].writingInProgress = false;
      }

to
      //set the writingInProgress flag to false
      for(var i in operations)
      {
        if (operations[i].key !== undefined)
        {
            db.buffer[operations[i].key].writingInProgress = false;
        }
      }

But I don't know what that could do if writingInProgress is not set.

**Comment by JohnMcLear** [comment]
I dunno tbh, I can't debug everyone's production environment for them sadly..

Have you run stress/load tests against your cluster?  Can you consistently replicate the problem?  Imho that's where I'd be starting then once you can replicate look at firing a fix in.

**Comment by ldidry** [comment]
Well, our stress tests is our users, and they are very good at it :sob:

After shutting down one big instance and restarting another, it works, but 2 big instances at once, it crashes. The culprit is the disk access time, slowing mongodb, and thus, etherpad.

Anyway, I think it could be useful to, at least, put a warning in logs if operations[i].key === undefined. We can let etherpad crashes if the admins can see a message like "It seems that your database doesn't respond correctly. It may be caused by timeout or slow disks" or anything else useful.

What do you think of that?

**Comment by JohnMcLear** [comment]
Yea that sounds fine :)

**Comment by ldidry** [comment]
Do you think my phrase is OK or do you want something else? Your choice, I'll do the pull request as soon as I know what to put in the warning.

**Comment by JohnMcLear** [comment]
Phrase looks fine :)  The real fix though is to put something in ueberdb that handles slow writes, correct?

**Comment by ldidry** [comment]
Yes, but I think it's beyond my power (I'm more a Perl guy)

**Comment by JohnMcLear** [comment]
Well @Pita maintains UeberDB but he has been kinda lazy at maintaining his projects lately and I don't have the time to look into this for you.   I suggest bumping @pita and creating an issue on ueberdb including a use case.

**Comment by ldidry** [comment]
I will, many thx.

For the record, I moved the VM to nodes with SSD disks and very low load (this is the only VM on these nodes), so there is no more timeout or slow answer. I get the same error, even with not visiting pads, just looking at a stat page (from [ep_pad_stats](https://www.npmjs.com/package/ep_pads_stats) plugin).

I had to apply my "if defined" patch to have etherpad stopping crashing. If I understand the code, writingInProgress is only used by the cache garbage collector, so my patch allows me to take the time to contact Pita.
